proxy.go

```diff
@@ -643,26 +643,68 @@ func (rp *reverseProxy) getUser(name string, password string) (found bool, u *us
 	defer rp.lock.RUnlock()
 	found = false
 	u = rp.users[name]
-	if u != nil {
+	switch {
+	case u != nil:
 		found = (u.password == password)
 		// existence of c and cu for toCluster is guaranteed by applyConfig
 		c = rp.clusters[u.toCluster]
 		cu = c.users[u.toUser]
-	} else if rp.hasWildcarded {
-		if s := strings.Split(name, "_"); len(s) == 2 {
-			u = rp.users[s[0]+"_*"]
-			if u != nil && u.isWildcarded {
-				c = rp.clusters[u.toCluster]
-				wildcardedCu := c.users[u.toUser]
-				if wildcardedCu != nil {
-					found = true
-					cu = wildcardedCu
-					cu.name = name
-					cu.password = password
+	case name == "" || name == defaultUser:
+		// default user can't work with the wildcarded feature for security reasons
+		found = false
+	case rp.hasWildcarded:
+		// checking if we have wildcarded users and if username matches one 3 possibles patterns
+		found, u, c, cu = findWildcardedUserInformation(rp, name, password)
+	}
+	return found, u, c, cu
+}
+
+func findWildcardedUserInformation(rp *reverseProxy, name string, password string) (found bool, u *user, c *cluster, cu *clusterUser) {
```

Comment:
I'd still make it a method receiver to reverseProxy, instead of a function accepting reverseProxy as parameter, but up to you the final choice

Reply:
done